### PR TITLE
*: Implement structured logging

### PR DESCRIPTION
**Pull request by derekparker**
Implements structured logging via Logrus. This gives us a logger per
boundry that we care about, allowing for easier parsing of logs if users
have more than one log option enabled. Also, cleans up a lot of
conditionals in the code by simply silencing the logger at creation as
opposed to conditionally logging everywhere.

### CODE REVIEW

File: pkg/dwarf/line/line_parser.go [hunk 1]
```diff
@@ -33,8 +36,7 @@ type DebugLineInfo struct {
 	// lastMachineCache[pc] is a state machine stopped at an address after pc
 	lastMachineCache map[uint64]*StateMachine
 
-	// logSuppressedErrors enables logging of otherwise suppressed errors
-	logSuppressedErrors bool
+	log *logrus.Entry
```
Comment (aarzilli) [inline pkg/dwarf/line/line_parser.go]: IMHO a reusable package like this shouldn't depend on a specific logging library that's not part of the standard library. Especially given the talk of contributing it to x/debug/dwarf.

File: service/rpccommon/server.go [hunk 2]
```diff
@@ -307,10 +296,8 @@ func (s *ServerImpl) serveJSONCodec(conn io.ReadWriteCloser) {
 		}
 
 		if mtype.Synchronous {
-			if s.log {
-				argvbytes, _ := json.Marshal(argv.Interface())
-				log.Printf("-> %s(%T%s)\n", req.ServiceMethod, argv.Interface(), argvbytes)
-			}
+			argvbytes, _ := json.Marshal(argv.Interface())
```
Comment (aarzilli) [inline service/rpccommon/server.go]: This is a fairly expensive operation, I don't think we should do it unconditionally given that rpc logging will mostly be disabled. Same goes for the other json.Marshal calls in this function and RPCCallback.Return.

File: pkg/dwarf/line/line_parser.go [hunk 3]
```diff
@@ -3,9 +3,12 @@ package line
 import (
 	"bytes"
 	"encoding/binary"
+	"io/ioutil"
 	"path/filepath"
 
 	"github.com/derekparker/delve/pkg/dwarf/util"
+	"github.com/derekparker/delve/pkg/logflags"
+	"github.com/sirupsen/logrus"
```
Comment (aarzilli) [inline pkg/dwarf/line/line_parser.go]: IMHO a reusable package like this shouldn't depend on a specific logging library that's not part of the standard library. Especially given the talk of contributing it to x/debug/dwarf.

Same goes for depending on logflags.

Comment (derekparker) [reply pkg/dwarf/line/line_parser.go]: All good points mentioned above, maybe we can change the package to accept a logger function, which will not add any imports.

Comment (aarzilli) [reply pkg/dwarf/line/line_parser.go]: I think that would be good.

**Comment by derekparker** [comment]
@aarzilli updated.

**Pull request merged**